### ISSUE TITLE
broken state save/restore since "store offset as opt arg for ggml_view_xd() operators"

### ISSUE THREAD

**Issue by niansa** [ggerganov/llama.cpp#1777]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [x] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

`llama_copy_state_data()` should do what it's documented to do

# Current Behavior

`llama_copy_state_data()` causes a Segmentation fault

# Environment and Context

* Physical (or virtual) hardware you are using, e.g. for Linux:

```
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         39 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  12
  On-line CPU(s) list:   0-11
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
    CPU family:          6
    Model:               158
    Thread(s) per core:  2
    Core(s) per socket:  6
    Socket(s):           1
    Stepping:            10
    CPU(s) scaling MHz:  17%
    CPU max MHz:         4700.0000
    CPU min MHz:         800.0000
    BogoMIPS:            7399.70
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts a
                         cpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch
                         _perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq 
                         dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_
                         2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowp
                         refetch cpuid_fault invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexprior
                         ity ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed a
                         dx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hw
                         p hwp_notify hwp_act_window hwp_epp md_clear flush_l1d arch_capabilities
Virtualization features: 
  Virtualization:        VT-x
Caches (sum of all):     
  L1d:                   192 KiB (6 instances)
  L1i:                   192 KiB (6 instances)
  L2:                    1.5 MiB (6 instances)
  L3:                    12 MiB (1 instance)
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-11
Vulnerabilities:         
  Itlb multihit:         KVM: Mitigation: Split huge pages
  L1tf:                  Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
  Mds:                   Mitigation; Clear CPU buffers; SMT vulnerable
  Meltdown:              Mitigation; PTI
  Mmio stale data:       Mitigation; Clear CPU buffers; SMT vulnerable
  Retbleed:              Mitigation; IBRS
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not a
                         ffected
  Srbds:                 Mitigation; Microcode
  Tsx async abort:       Mitigation; TSX disabled
```

* Operating System, e.g. for Linux:

```
Linux lyllia 6.1.27 #3 SMP PREEMPT_DYNAMIC Thu Jun  8 14:05:39 CEST 2023 x86_64 GNU/Linux
```

* SDK version, e.g. for Linux:

```
g++ (Debian 12.2.0-14) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

(The rest is not applicable here)

# Failure Information (for bugs)

This regression was introduced in https://github.com/ggerganov/llama.cpp/pull/1642/commits/94ea9e7bfecc1116afc830606cf886f1b3620257

# Steps to reproduce

Just call `llama_copy_state_data()` the way you normally would. Nothing too complicated here.

# Failure Logs

```
#0  0x00007ffff6e16d34 in ggml_view_3d (ctx=0x7ffff6eeebd8 <g_state+200>, a=0x7ffed47ff030, ne0=<optimized out>, ne1=<optimized out>, ne2=32, nb1=8192, nb2=16777216, offset=0) at ggml.c:5901
        ne = {4096, 23, 32, 1}
        is_node = <optimized out>
        result = 0x7ffffffe50d0
        offs = 0x7ffffffe51d0

#1  0x00007ffff6e00a69 in llama_copy_state_data (ctx=0x55555556e1d0, dst=<optimized out>) at llama.cpp:2751
        buffer = <uninitialized>
        kout3d = <optimized out>
        v3d = <optimized out>
        gf = <error reading variable gf (value of type `ggml_cgraph' requires 98360 bytes, which is more than max-value-size)>
        elt_size = 2
        cpy_ctx = 0x7ffff6eeebd8 <g_state+200>
        vout3d = <optimized out>
        k3d = <optimized out>
        kv_size = <optimized out>
        kv_self = @0x55555556f5f8: {k = 0x7ffed47ff030, v = 0x7ffef47ff130, ctx = 0x7ffff6eeeb78 <g_state+104>, buf = {addr = 0x7ffed47ff010 " ", size = 1075838976}, n = 23}
        hparams = @0x55555556f59c: {n_vocab = 32000, n_ctx = 2048, n_embd = 4096, n_mult = 256, n_head = 32, n_layer = 32, n_rot = 128, ftype = LLAMA_FTYPE_MOSTLY_Q4_0}
        kv_ntok = <optimized out>
        n_layer = <optimized out>
        n_embd = <optimized out>
        n_ctx = <optimized out>
        out = 0x7ffe2397843c ""
        written = <optimized out>
        max_size = <optimized out>

#2  0x00007ffff7fba1f1 in (anonymous namespace)::CppAPITest_llama_7B_ggmlv3_StateSave_Test::TestBody (this=this@entry=0x55555578a5c0) at ../llmodel_cpp_test.cpp:57
        gtest_ar = {success_ = 192, message_ = {_M_t = {<std::__uniq_ptr_impl<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::default_delete<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >> = {_M_t = {<std::_Tuple_impl<0ul, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::default_delete<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >> = {<std::_Tuple_impl<1ul, std::default_delete<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >> = {<std::_Head_base<1ul, std::default_delete<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, true>> = {_M_head_impl = {<No data fields>}}, <No data fields>}, <std::_Head_base<0ul, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, false>> = {_M_head_impl = 0x55555578a5c0}, <No data fields>}, <No data fields>}}, <No data fields>}}}
```

**Comment by ggerganov** [comment]
I believe this is fixed with 17c10acfb44ecb7af25e37fb67b9501cbc0034d2

Let me know

**Comment by niansa** [comment]
That seems very much related, yes! Will check asap
